File: local-ai/src-tauri/src/commands/memory.rs
use tauri::State;

use crate::services::agent_repo::AgentRepository;
use crate::services::memory::MemoryService;
use crate::types::{CuratorPackage, DailyLog, MemoryContext, MemoryFile, MessageAttachment};
use crate::DatabaseState;

pub struct MemoryState {
    pub root_path: String,
}

impl MemoryState {
    fn resolve_service(&self, db: &crate::services::database::Database) -> Result<MemoryService, String> {
        let agent_repo = AgentRepository::new(db);
        let workspace_path = agent_repo.resolve_active_workspace_path(&self.root_path)?;
        let service = MemoryService::new(&workspace_path);
        service.initialize()?;
        Ok(service)
    }
}

#[tauri::command]
pub async fn memory_initialize(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<(), String> {
    let service = state.resolve_service(&db_state.db)?;
    service.initialize()
}

#[tauri::command]
pub async fn memory_read_file(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    filename: String,
) -> Result<MemoryFile, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.read_file(&filename)
}

#[tauri::command]
pub async fn memory_write_file(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    filename: String,
    content: String,
) -> Result<(), String> {
    let service = state.resolve_service(&db_state.db)?;
    service.write_file(&filename, &content)
}

#[tauri::command]
pub async fn memory_get_today_log(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<DailyLog, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.get_today_log()
}

#[tauri::command]
pub async fn memory_append_log(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    entry: String,
) -> Result<(), String> {
    let service = state.resolve_service(&db_state.db)?;
    service.append_to_today_log(&entry)
}

#[tauri::command]
pub async fn memory_list_daily_logs(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<Vec<String>, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.list_daily_logs()
}

#[tauri::command]
pub async fn memory_list_knowledge(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<Vec<String>, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.list_knowledge_files()
}

#[tauri::command]
pub async fn memory_list_curator_staged(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<Vec<CuratorPackage>, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.list_curator_staged_packages()
}

#[tauri::command]
pub async fn memory_import_curator_package(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    folder_name: String,
) -> Result<String, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.import_curator_package(&folder_name)
}

#[tauri::command]
pub async fn memory_reject_curator_package(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    folder_name: String,
) -> Result<(), String> {
    let service = state.resolve_service(&db_state.db)?;
    service.reject_curator_package(&folder_name)
}

#[tauri::command]
pub async fn memory_load_context(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<MemoryContext, String> {
    let service = state.resolve_service(&db_state.db)?;
    service.load_context()
}

#[tauri::command]
pub async fn memory_get_base_path(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<String, String> {
    let service = state.resolve_service(&db_state.db)?;
    Ok(service.get_base_path())
}

#[tauri::command]
pub async fn memory_open_folder(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
) -> Result<(), String> {
    let service = state.resolve_service(&db_state.db)?;
    service.open_base_path()
}

#[tauri::command]
#[allow(non_snake_case)]
pub async fn memory_store_chat_attachment(
    db_state: State<'_, DatabaseState>,
    state: State<'_, MemoryState>,
    conversationId: String,
    filename: String,
    kind: String,
    mimeType: Option<String>,
    bytes: Vec<u8>,
) -> Result<MessageAttachment, String> {
    if kind != "image" && kind != "audio" {
        return Err("Unsupported attachment kind".to_string());
    }

    let service = state.resolve_service(&db_state.db)?;
    service.store_chat_attachment(
        &conversationId,
        &filename,
        &kind,
        mimeType.as_deref(),
        &bytes,
    )
}

File: local-ai/src/services/memory.ts
import { invoke } from '@tauri-apps/api/core';
import type { CuratorPackage } from '@/types';

interface RawCuratorPackage {
  id: string;
  folder_name: string;
  title: string;
  summary?: string;
  source?: string;
  tags: string[];
  request_topic?: string;
  created_at?: string;
  status: string;
  path: string;
}

export interface MemoryFile {
  name: string;
  path: string;
  content: string;
  exists: boolean;
  modified_at?: string;
}

export interface DailyLog {
  date: string;
  path: string;
  content: string;
  exists: boolean;
}

export interface MemoryContext {
  soul?: string;
  user?: string;
  memory?: string;
  today_log?: string;
  knowledge_files: string[];
}

function mapCuratorPackage(pkg: RawCuratorPackage): CuratorPackage {
  return {
    id: pkg.id,
    folderName: pkg.folder_name,
    title: pkg.title,
    summary: pkg.summary,
    source: pkg.source,
    tags: pkg.tags,
    requestTopic: pkg.request_topic,
    createdAt: pkg.created_at,
    status: pkg.status,
    path: pkg.path,
  };
}

export const memoryApi = {
  async initialize(): Promise<void> {
    return invoke('memory_initialize');
  },

  async readFile(filename: string): Promise<MemoryFile> {
    return invoke('memory_read_file', { filename });
  },

  async writeFile(filename: string, content: string): Promise<void> {
    return invoke('memory_write_file', { filename, content });
  },

  async getTodayLog(): Promise<DailyLog> {
    return invoke('memory_get_today_log');
  },

  async appendLog(entry: string): Promise<void> {
    return invoke('memory_append_log', { entry });
  },

  async listDailyLogs(): Promise<string[]> {
    return invoke('memory_list_daily_logs');
  },

  async listKnowledge(): Promise<string[]> {
    return invoke('memory_list_knowledge');
  },

  async listCuratorStaged(): Promise<CuratorPackage[]> {
    const packages = await invoke<RawCuratorPackage[]>('memory_list_curator_staged');
    return packages.map(mapCuratorPackage);
  },

  async importCuratorPackage(folderName: string): Promise<string> {
    return invoke('memory_import_curator_package', { folderName });
  },

  async rejectCuratorPackage(folderName: string): Promise<void> {
    return invoke('memory_reject_curator_package', { folderName });
  },

  async loadContext(): Promise<MemoryContext> {
    return invoke('memory_load_context');
  },

  async getBasePath(): Promise<string> {
    return invoke('memory_get_base_path');
  },

  async openFolder(): Promise<void> {
    return invoke('memory_open_folder');
  },
};

